### PR TITLE
feat: add upgrade section in workspace settings

### LINKED ISSUE #2727: Feature: Allow users to upgrade workspaces
### Type of feature

🍕 Feature

### Current behavior

Currently, users can only upgrade their individual accounts under user settings.

### Suggested solution

Each workspace will have an upgrade section to allow paid features like setting it to public or private. The subscriptions will be individualized and be paid using a Stripe checkout. Needs https://github.com/open-sauced/api/pull/576 to be implemented.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by zeucapua**
## Description

Implementing a section that will allow users to upgrade the workspace.
- [x] Add designed section
- [x] Change section to manage subscription if in a paid workspace
- [x] Go to Stripe checkout on click

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Closes #2727 
Needs https://github.com/open-sauced/api/pull/576

## Mobile & Desktop Screenshots/Recordings

I believe Stripe redirects to the beta app, and the manage subscription works, I just had to go to the other tab.

https://github.com/open-sauced/app/assets/20603494/1bfd5869-96fb-450e-8082-3e0b065c52a2

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

upgrade-you-beyonce-gif

### CODE REVIEW

**Comment by isabensusan** [comment]
this is looking great!!

I know this is only a draft, but just in case, here's a couple notes on styling

Font fixes:
image

Padding fixes:
image

thank youuuu

File: lib/utils/workspace-utils.ts [hunk 1]
```diff
@@ -76,6 +76,17 @@ export async function saveWorkspace({
   return { data: { workspace: data, repos: repoData, contributors: contributorsData }, error };
 }
 
+export async function upgradeWorkspace({ workspaceId, sessionToken }: { workspaceId: string; sessionToken: string }) {
+  const { data, error } = await fetchApiData<{ sessionId: string }>({
+    path: `auth/checkout/workspaces/${workspaceId}/session`,
+    method: "POST",
+    body: {},
```
Comment (jpmcb) [inline lib/utils/workspace-utils.ts:83]: If this isn't getting added automatically by the empty post request, a header here for `Content-Length: 0` would be ideal

**Comment by isabensusan** [comment]
Thought you might need this too -- once the workspace is upgraded we should show a section to let the user cancel  their pro workspace subscription. Quick and dirty sketch:

image


image

**Comment by jpmcb** [comment]
> Is there a cancel subscription endpoint I should be hitting? I assume there must be a Stripe thing in the backend to ensure they don't get recurring charges after cancelling @jpmcb

Yeah, good question: the webhook will indeed handle canceling the subscriptions (so if the customer logs into their stripe portal and manually cancels it, that gets handled by the webhook event from stripe). We could do something similar where we have a cancel button that forwards them to their stripe dashboard via the `NEXT_PUBLIC_STRIPE_SUB_CANCEL_URL` env var like we currently do for "pro" plans:

https://github.com/open-sauced/app/blob/7e7c3be849b02c44987e293c4e9c0a4c26b9f534/components/organisms/UserSettingsPage/user-settings-page.tsx#L471-L479

Testing this out, it indeed takes me to a dashboard where I can cancel the subscription:

Screenshot 2024-02-22 at 11 49 09 AM

I say we do ^ for now since it'll be easy.

But it's not a great user experience since it'd be confusing for anyone paying for multiple workspaces what they're doing: a few questions for @isabensusan - should the OpenSauced platform handle canceling the subscription? Or should we forward them onto stripe? We could add an endpoint in the API and handle it on our side, just depends on how we want to go about it and what the user experience should feel like.

**Comment by isabensusan** [comment]
@jpmcb for now, I'd leave it on the Stripe side -- but we do need an entry point on the website for users to get to the stripe page and be able to cancel their subscriptions

File: .env [hunk 2]
```diff
@@ -4,7 +4,7 @@ NEXT_PUBLIC_API_URL=https://beta.api.opensauced.pizza/v2
 NEXT_PUBLIC_EXP_API_URL=https://beta.api.opensauced.pizza/v2
 NEXT_PUBLIC_POSTHOG_ID=phc_Y0xz6nK55MEwWjobJsI2P8rsiomZJ6eZLoXehmMy9tt
 NEXT_PUBLIC_CLOUD_NAME=dgxgziswe
-NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY=test
+NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY=pk_test_51H1yp3JnZ7CYWRG2KmQMIE8sizoaSaZ8FxlEsc5uQb81gVswDlWu6FErmcukqc6Hso3mhCWgnLsjWphight4t1Mf00jZZjckyZ
```
Comment (brandonroberts) [inline .env:7]: Revert this as the test key is applied when deployed

File: pages/workspaces/[workspaceId]/settings.tsx [hunk 3]
```diff
@@ -1,9 +1,12 @@
+import { FaRegCheckCircle } from "react-icons/fa";
 import { useRouter } from "next/router";
 import { GetServerSidePropsContext } from "next";
 import { createPagesServerClient } from "@supabase/auth-helpers-nextjs";
 import { ComponentProps, useState } from "react";
 import dynamic from "next/dynamic";
 import { useEffectOnce } from "react-use";
+import { loadStripe } from "@stripe/stripe-js";
```
Comment (brandonroberts) [inline pages/workspaces/[workspaceId]/settings.tsx:8]: We have a utility already for loading stripe in `lib/utils/stripe-client.ts`

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Review by brandonroberts — APPROVED** [review]

**Review by jpmcb — APPROVED** [review]

**Pull request merged**